File: src/factoriorl/knowledge.py
"""What the game *is*, as opposed to what the world currently looks like.

The mod has always been able to read `prototypes.*` and has never returned any
of it. Three call sites in `actions.lua` look something up -- a recipe's first
product, an item's `place_result` -- and throw it away as soon as the action
answers. The consequences were paid elsewhere: a client had no way to learn that
an iron gear wheel costs two plates except by attempting the craft, and
`tasks/spec.py` carries a hardcoded `ENTITY_TILE_SIZES` table of four prototypes
with a 1x1 fallback for everything else, because the footprint of a stone
furnace was not askable. `mod/factoriorl/knowledge.lua` returns it; this module
fetches, caches and renders it.

Why it is cached on disk
------------------------
The reply is the largest the protocol produces and it does not change while a
worker runs. Fetching it once per run and reusing the file is worth a little
machinery, but only if the key is honest about what can invalidate it: the
**engine build**, because prototypes come from the game, and
`manifest.mod_source_digest()`, because a change to `knowledge.lua` changes what
is returned from the same game. Keying on the build alone would serve a stale
file after any edit to the mod, and the difference would be invisible -- the
JSON would parse and simply be missing a field.

Why `render` must be byte-stable
--------------------------------
Its output goes into a prompt prefix that the provider caches. A reordered line
is indistinguishable from a changed one to a cache key, so a single `pairs`
iteration leaking through would invalidate the entire cached conversation on
every turn and be visible only as a bill. `knowledge.lua` sorts each collection
before it serialises; this module sorts again on the way out rather than trust
that, because a cache file written by an older mod is exactly the input nobody
would think to check.
"""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any

from factoriorl.manifest import mod_source_digest
from factoriorl.paths import runtime_dir

#: The three tables `knowledge.lua` returns. Each is required to be non-empty:
#: no Factorio install has zero recipes, so an empty one means the Lua walked a
#: table it could not read and returned success anyway.
SECTIONS = ("recipes", "placeable", "technologies")

#: Cache filenames are derived from an engine build string, which comes from the
#: engine rather than from this repository. Anything outside this set is
#: replaced rather than trusted into a path.
_UNSAFE = re.compile(r"[^A-Za-z0-9._-]")


class KnowledgeUnavailable(RuntimeError):
    """The worker could not supply the static game data.

    Raised rather than defaulted, for the reason `freeplay.FreeplayUnreadable`
    gives: a fabricated recipe graph would make a run look like it was planning
    against the real game while it was not, and no later measurement could tell.
    """


# --------------------------------------------------------------- fetching


def fetch(session: Any) -> dict:
    """One `knowledge` request against a live worker.

    Read-only, so it needs no episode and can be asked before the first reset.
    """
    timed = session.knowledge()
    response = timed.response
    if not response.ok:
        detail = response.error.message if response.error else response.code
        raise KnowledgeUnavailable(f"the worker refused the knowledge request: {detail}")
    result = response.result or {}
    missing = [section for section in SECTIONS if not _rows(result, section)]
    if missing:
        raise KnowledgeUnavailable(
            f"the knowledge reply has no {', '.join(missing)}. Every Factorio install "
            f"has hundreds of each, so an empty table is a read that silently failed "
            f"rather than a game with nothing in it"
        )
    return result


# ---------------------------------------------------------------- caching


def cache_key(engine_build: str, mod_digest: str | None = None) -> str:
    """What a cached answer is allowed to be reused for.

    Both halves matter. The prototypes come from the engine, so a different
    build is a different answer; and `knowledge.lua` decides which fields are
    reported, so a mod edit is a different answer from the same engine.
    """
    digest = mod_digest if mod_digest is not None else mod_source_digest()
    return f"{_UNSAFE.sub('-', engine_build or 'unknown')}-{_UNSAFE.sub('-', digest)}"


def cache_path(engine_build: str, mod_digest: str | None = None) -> Path:
    return runtime_dir() / "knowledge" / f"{cache_key(engine_build, mod_digest)}.json"


def cached(session: Any, engine_build: str, mod_digest: str | None = None) -> dict:
    """Fetch once per (engine build, mod digest) and reuse the file thereafter.

    A cache miss, an unreadable file and a file whose stored key disagrees with
    the one asked for all mean the same thing: fetch. The key is stored inside
    the file as well as in its name because the name is slugged, so two builds
    whose strings differ only in a character this module rewrites would
    otherwise share one entry.
    """
    key = cache_key(engine_build, mod_digest)
    path = cache_path(engine_build, mod_digest)
    stored = _read_cache(path, key)
    if stored is not None:
        return stored

    knowledge = fetch(session)
    document = {
        "key": key,
        "engine_build": engine_build,
        "mod_digest": mod_digest if mod_digest is not None else mod_source_digest(),
        "knowledge": knowledge,
    }
    try:
        path.parent.mkdir(parents=True, exist_ok=True)
        # Written whole and moved into place: a crash mid-write would otherwise
        # leave a truncated file that the next run has to recognise and repair.
        temporary = path.with_suffix(".json.tmp")
        temporary.write_text(json.dumps(document, indent=2, sort_keys=True), encoding="utf-8")
        temporary.replace(path)
    except OSError:
        # Failing to cache is not failing to fetch. The caller has its answer.
        pass
    return knowledge


def _read_cache(path: Path, key: str) -> dict | None:
    try:
        document = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError):
        return None
    if not isinstance(document, dict) or document.get("key") != key:
        return None
    knowledge = document.get("knowledge")
    if not isinstance(knowledge, dict):
        return None
    if any(not _rows(knowledge, section) for section in SECTIONS):
        return None
    return knowledge


# --------------------------------------------------------------- rendering


def render(knowledge: dict) -> str:
    """A prompt-sized plain-text block. Byte-identical for equal input.

    Deterministic by construction: every collection is sorted by name here, not
    only in Lua, and every number goes through `_number` so a float that arrives
    as `0.5` and one that arrives as `0.5000000001` cannot render differently on
    two runs of the same worker.
    """
    lines = [
        "=== STATIC GAME KNOWLEDGE ===",
        "Factorio prototype data: what things cost, what they make, how big they",
        "are, and what unlocks what. The same on every turn, and nothing you do",
        "changes it.",
        "",
        "It says nothing about what is available to you *now*. The recipes you can",
        "currently craft are in the observation under RECIPES, and the research you",
        "can start is under ARGUMENT VALUES. Read availability there, not here.",
        "",
    ]
    lines += _recipe_section(knowledge)
    lines.append("")
    lines += _placeable_section(knowledge)
    lines.append("")
    lines += _technology_section(knowledge)
    return "\n".join(lines)


def _recipe_section(knowledge: dict) -> list[str]:
    rows = _rows(knowledge, "recipes")
    lines = [f"RECIPES ({len(rows)}) -- name: ingredients -> products (craft seconds, category)"]
    for row in rows:
        name = str(row.get("name", ""))
        ingredients = _stacks(row.get("ingredients")) or "-"
        # The product name is dropped when it repeats the recipe name, which is
        # the overwhelming majority of recipes and is the difference between a
        # line the model reads once and a line it reads twice.
        products = _stacks(row.get("products"), omit=name) or "-"
        line = (
            f"{name}: {ingredients} -> {products} "
            f"({_number(row.get('energy'))}s, {row.get('category', 'crafting')})"
        )
        # No lock marker here, and none for entities or technologies below.
        # `enabled`, `unlocked` and `researched` are force state read off
        # `game.forces["player"]`, not prototype state, and they move the
        # moment research completes. This block is captured once and pinned
        # into a prompt prefix that must stay byte-identical for the whole
        # run, so a marker recorded here would be quietly wrong for most of
        # it. Availability is published every turn in the observation --
        # enabled recipes under RECIPES, the researchable frontier under
        # ARGUMENT VALUES -- which is where a reader should look for it.
        if row.get("hidden"):
            # Still craftable by a machine; only kept out of the crafting menu.
            line += " [hidden]"
        lines.append(line)
    return lines


def _placeable_section(knowledge: dict) -> list[str]:
    rows = _rows(knowledge, "placeable")
    lines = [
        f"PLACEABLE ENTITIES ({len(rows)}) -- item: tiles wide x tall, "
        f"footprint centred on the placement position"
    ]
    for row in rows:
        name = str(row.get("name", ""))
        line = f"{name}: {_number(row.get('width', 1))}x{_number(row.get('height', 1))}"
        entity = row.get("entity")
        # Named only when it differs. `place` takes the item name, so repeating
        # an identical entity name on every line would be noise the agent has to
        # read past on all ~200 of them.
        if entity and entity != name:
            line += f" -> {entity}"
        # The stricter rule `actions.lua` applies: placement is refused while a
        # same-named recipe is locked, which base Factorio does not do.
        lines.append(line)
    return lines


def _technology_section(knowledge: dict) -> list[str]:
    rows = _rows(knowledge, "technologies")
    lines = [f"TECHNOLOGIES ({len(rows)}) -- name: cost -> recipes unlocked <- prerequisites"]
    for row in rows:
        name = str(row.get("name", ""))
        line = f"{name}: {_research_cost(row)}"
        unlocks = sorted(str(entry) for entry in _list(row.get("unlocks")))
        if unlocks:
            line += " -> " + ", ".join(unlocks)
        prerequisites = sorted(str(entry) for entry in _list(row.get("prerequisites")))
        if prerequisites:
            line += " <- " + ", ".join(prerequisites)
        lines.append(line)
    return lines


def _research_cost(row: dict) -> str:
    """The cost line for one technology, including the ones that have no cost.

    A trigger technology is reported as its trigger rather than as a science
    cost, because it genuinely has none: 2.0 completes 7 of 196 by crafting or
    mining an item, and they cannot be queued at all. `protocol.py` carries
    `tech_not_selectable` for exactly this case, and an agent told only "not
    selectable" has nothing to act on.
    """
    trigger = row.get("trigger")
    if trigger:
        # What actually satisfies it. `trigger:craft-item` told an agent it
        # could not research the thing and nothing about how to obtain it, which
        # is the half that matters -- these unlock by *doing* something, and the
        # something is knowable.
        subject = row.get("trigger_item") or row.get("trigger_entity")
        count = row.get("trigger_count")
        verb = str(trigger).replace("-", " ")
        if subject and count:
            return f"no science: {verb} {_number(count)}x {subject}"
        if subject:
            return f"no science: {verb} {subject}"
        return f"no science: completes by {verb}"
    count = row.get("unit_count")
    ingredients = _list(row.get("unit_ingredients"))
    if not ingredients:
        return "free" if count is None else f"{_number(count)}x -"
    packs = "+".join(_stack(entry, brief=True) for entry in ingredients)
    return f"{_number(count) if count is not None else '?'}x {packs}"


# ----------------------------------------------------------------- helpers


def _rows(knowledge: dict, section: str) -> list[dict]:
    """One of the three tables, name-sorted, whatever shape the JSON arrived in.

    Lua's empty table serialises as a JSON *object*, not an array, so an install
    with nothing in a section hands back `{}` where every other reply hands back
    a list. Normalised here rather than at each of the five call sites.
    """
    value = knowledge.get(section)
    if isinstance(value, dict):
        value = list(value.values())
    if not isinstance(value, list):
        return []
    rows = [row for row in value if isinstance(row, dict)]
    return sorted(rows, key=lambda row: str(row.get("name", "")))


def _list(value: Any) -> list:
    if isinstance(value, dict):
        return list(value.values())
    return list(value) if isinstance(value, list) else []


def _stacks(entries: Any, omit: str | None = None) -> str:
    return " + ".join(
        _stack(entry, omit=omit) for entry in _list(entries) if isinstance(entry, dict)
    )


def _stack(entry: dict, omit: str | None = None, brief: bool = False) -> str:
    """One ingredient or product: `2 iron-plate`, and the awkward cases.

    A probabilistic product carries a range and a chance instead of an amount --
    uranium processing, every ore-crushing recipe -- and rendering the missing
    `amount` as nothing would read as a recipe that produces one of something.
    """
    name = str(entry.get("name", ""))
    if brief:
        return name
    amount = entry.get("amount")
    low, high = entry.get("amount_min"), entry.get("amount_max")
    if low is not None and high is not None and low != high:
        rendered = f"{_number(low)}-{_number(high)}"
    else:
        rendered = _number(amount)
    label = name
    if entry.get("type") == "fluid":
        label += "(fluid)"
    probability = entry.get("probability")
    if isinstance(probability, int | float) and probability < 1:
        label += f"@{_number(probability * 100)}%"
    # The product of a recipe named after it: `iron-gear-wheel: 2 iron-plate -> 1`.
    # Only when the label is the bare name -- a fluid or a chance is information
    # the recipe's own name does not carry, so it stays.
    if omit is not None and label == omit:
        return rendered
    return f"{rendered} {label}"


def _number(value: Any) -> str:
    """Numbers that render the same way twice.

    Whole floats lose their `.0` -- `2.0 iron-plate` is a token spent on
    nothing -- and the rest are rounded before formatting, because recipe
    energies arrive from the engine as floats and `0.30000000000000004` would be
    both unreadable and a different string on a machine that rounded elsewhere.
    """
    if value is None:
        return "?"
    if isinstance(value, bool) or not isinstance(value, int | float):
        return str(value)
    rounded = round(float(value), 4)
    if rounded == int(rounded):
        return str(int(rounded))
    return f"{rounded:g}"

File: mod/factoriorl/knowledge.lua
-- Static game data, read straight out of the loaded prototypes.
--
-- The mod has always had `prototypes.*` in reach and has never returned any of
-- it. Three call sites use it, all of them to *check* something --
-- `actions.lua` reads `prototypes.recipe[...].products[1]` to know what a craft
-- will produce, `prototypes.item[...].place_result` to know whether an item
-- places anything -- and every one of those facts is thrown away as soon as the
-- action answers. So a client that wants to plan has no way to learn what iron
-- gear wheels are made of except by trying to craft them, and
-- `tasks/spec.py` ended up carrying a hardcoded `ENTITY_TILE_SIZES` table of
-- four prototypes because the footprint of a stone furnace was not askable.
--
-- Read-only, and deliberately outside `MUTATING`: this reads prototypes and
-- force state and writes nothing, so deduplicating it would make a second
-- request return a stored answer instead of a current one.
--
-- What is *not* static
-- --------------------
-- Recipe `enabled`, technology `researched` and the placement gate all come off
-- `game.forces["player"]`, not off a prototype: they are the force's state at
-- the moment of the call, and research moves them. The reply says which tick it
-- was taken at so a caller caching this on disk can tell how old the mutable
-- half is. Everything else -- ingredients, products, energy, tile sizes,
-- prerequisites, unlock effects -- is fixed for a given engine build and mod
-- set.
--
-- Ordering is settled here rather than left to the caller. Every collection
-- below is built by walking a Lua hash (`force.recipes`, `prototypes.item`,
-- `force.technologies`), and `pairs` order is not stable across runs, so an
-- unsorted reply would serialise differently every time it was fetched. The
-- Python side puts this in a cached prompt prefix, where a reordered line is
-- indistinguishable from a changed one.

local knowledge = {}

local function player_force()
  return game.forces["player"]
end

local function sorted_keys(collection)
  local names = {}
  for name in pairs(collection) do names[#names + 1] = name end
  table.sort(names)
  return names
end

--- `{ name = ..., amount = ... }` for a recipe's ingredient or product list.
--
-- Kept in prototype order, which is an array and therefore already stable --
-- sorting it would scramble the reading order a recipe is written in.
--
-- Two shapes need flattening. A fluid ingredient is the same table with
-- `type = "fluid"`, and that distinction matters to an agent because a fluid
-- cannot be carried or hand-crafted; it is reported only when it is not
-- "item", so the common case costs nothing. A probabilistic product carries
-- `amount_min`/`amount_max`/`probability` instead of `amount` -- uranium
-- processing and every ore-crushing recipe do -- and reporting `nil` there
-- would render as a missing number rather than as a range.
local function stack_list(entries)
  local out = {}
  for index, entry in ipairs(entries or {}) do
    local stack = { name = entry.name, amount = entry.amount }
    if stack.amount == nil then
      stack.amount = entry.amount_max or entry.amount_min or 0
      stack.amount_min = entry.amount_min
      stack.amount_max = entry.amount_max
    end
    if entry.type and entry.type ~= "item" then stack.type = entry.type end
    -- Only when it is actually a chance. `probability = 1` is the default and
    -- would otherwise appear on every product of every recipe.
    if entry.probability and entry.probability < 1 then
      stack.probability = entry.probability
    end
    out[index] = stack
  end
  return out
end

--- Every recipe the player force knows about, enabled or not.
--
-- Read off `force.recipes` rather than `prototypes.recipe` because the two
-- carry different halves of the answer: the prototype has the ingredients and
-- the timing, the force has whether it is unlocked, and a planner needs both in
-- the same row. `force.recipes` is keyed by every recipe prototype, so nothing
-- is lost by going through it.
--
-- Recipes with no products are dropped. That is not a taste judgement: on
-- 2.0.60 `parameter-0` through `parameter-9` are enabled, unhidden, and have
-- zero ingredients and zero products -- placeholders for parametrised
-- blueprints. `observations.lua` already filters them on the same property,
-- where they had been occupying 10 of the 22 recipe slots a policy could see.
-- The count that was dropped is reported, so the omission is visible rather
-- than inferred from a short list.
function knowledge.recipes()
  local force = player_force()
  local out, dropped = {}, 0
  if not force then return out, dropped end
  for _, name in ipairs(sorted_keys(force.recipes)) do
    local recipe = force.recipes[name]
    if not recipe.products or #recipe.products == 0 then
      dropped = dropped + 1
    else
      out[#out + 1] = {
        name = name,
        ingredients = stack_list(recipe.ingredients),
        products = stack_list(recipe.products),
        -- Seconds at crafting speed 1, which is what the recipe tooltip shows
        -- and what `actions.lua` already returns from a started craft.
        energy = recipe.energy,
        category = recipe.category,
        -- Force state, not prototype state. See the header.
        enabled = recipe.enabled,
        -- Reported only when true, so the common row stays short. A hidden
        -- recipe still runs -- it is only kept out of the crafting menu -- so
        -- dropping these would hide real production routes.
        hidden = recipe.hidden or nil,
      }
    end
  end
  return out, dropped
end

--- Items that place an entity, with the footprint that placement needs.
--
-- Keyed by the **item** name, because that is what the `place` action takes and
-- what the agent holds in its inventory. The entity name is carried alongside
-- and is usually but not always the same string.
--
-- `tile_width`/`tile_height` are the numbers `tasks/spec.py` hardcodes for four
-- prototypes and guesses as 1x1 for everything else. A guessed footprint is
-- worse than no footprint: it decides where a task believes an entity fits.
--
-- The recipe gate is reported because `actions.lua` applies a stricter rule
-- than the base game does. Base Factorio gates technology at the recipe layer
-- only, so holding an item is normally enough to place it; the action profile
-- refuses placement when a recipe of the same name exists and is disabled, so
-- PLAN 2.1's "placement cannot bypass technology restrictions" is directly
-- testable. An agent cannot predict that refusal without knowing which item
-- names have a recipe behind them.
function knowledge.placeable()
  local force = player_force()
  local out = {}
  for _, name in ipairs(sorted_keys(prototypes.item)) do
    local item = prototypes.item[name]
    local entity = item.place_result
    if entity then
      local gate = force and force.recipes[name] or nil
      out[#out + 1] = {
        name = name,
        entity = entity.name,
        width = entity.tile_width,
        height = entity.tile_height,
        -- Named rather than boolean: an agent that is refused a placement needs
        -- to know which recipe to research, and the item name is not always the
        -- recipe name for other prototypes.
        gated_by = gate and gate.name or nil,
        -- Force state, like `recipes[].enabled`. Absent when nothing gates it.
        unlocked = gate and gate.enabled or nil,
      }
    end
  end
  return out
end

--- The technology tree: cost, prerequisites, and what each unlock buys.
--
-- `unlocks` is the unlock-recipe effects only. The others (bonuses, inserter
-- capacity, character reach) do not name a recipe, and the question this table
-- exists to answer is "what do I research to be allowed to craft X".
--
-- `trigger` exists because Factorio 2.0 roots the tree in trigger technologies:
-- 7 of 196 complete by crafting or mining an item and cannot be queued at all,
-- which is why `protocol.py` carries `tech_not_selectable` as a code distinct
-- from `tech_locked`. Without this field an agent told "not selectable" has no
-- way to learn what would actually finish it, and would wait for prerequisites
-- that never arrive.
--
-- `research_unit_count` is behind a pcall: for an infinite technology it is
-- derived from a formula and a level rather than stored, and raising here would
-- take the whole request down over one late-game entry.
function knowledge.technologies()
  local force = player_force()
  local out = {}
  if not force then return out end
  for _, name in ipairs(sorted_keys(force.technologies)) do
    local tech = force.technologies[name]
    local prerequisites = sorted_keys(tech.prerequisites or {})

    local unlocks = {}
    for _, effect in ipairs(tech.prototype.effects or {}) do
      if effect.type == "unlock-recipe" then unlocks[#unlocks + 1] = effect.recipe end
    end
    table.sort(unlocks)

    local ok, count = pcall(function() return tech.research_unit_count end)
    local trigger = tech.prototype.research_trigger

    out[#out + 1] = {
      name = name,
      prerequisites = prerequisites,
      unit_count = ok and count or nil,
      unit_ingredients = stack_list(tech.research_unit_ingredients),
      unlocks = unlocks,
      trigger = trigger and trigger.type or nil,
      -- Force state, like `recipes[].enabled`. See the header.
      researched = tech.researched,
    }
  end
  return out
end

--- Everything above, in one reply.
--
-- Counts are returned beside the tables so a caller can check that a truncated
-- or partially serialised answer is not silently accepted as a complete one --
-- this reply is the largest the protocol produces, and it is cached to disk and
-- then reused for the rest of a run.
function knowledge.snapshot()
  local recipes, dropped = knowledge.recipes()
  local placeable = knowledge.placeable()
  local technologies = knowledge.technologies()
  return {
    recipes = recipes,
    placeable = placeable,
    technologies = technologies,
    counts = {
      recipes = #recipes,
      placeable = #placeable,
      technologies = #technologies,
      -- Product-less placeholder recipes. See `knowledge.recipes`.
      recipes_without_products = dropped,
    },
    -- The mutable half -- `enabled`, `unlocked`, `researched` -- is the force's
    -- state at this tick and nothing else. Stated in the reply because the
    -- Python side caches this keyed on engine build and mod digest, neither of
    -- which moves when a technology completes.
    force_state_tick = game.tick,
    force = "player",
  }
end

return knowledge
